File: 189_Rotate_Array.py
class Solution:
    def rotate(self, nums: list[int], k: int) -> None:
        n = len(nums)
        k %= n  # Handle cases where k is greater than n

        def reverse(start: int, end: int) -> None:
            while start < end:
                nums[start], nums[end] = nums[end], nums[start]
                start += 1
                end -= 1
    
        reverse(0, n - 1)   # Step 1: Reverse the entire array  
        reverse(0, k - 1)   # Step 2: Reverse the first k elements
        reverse(k, n - 1)   # Step 3: Reverse the remaining n-k elements

        return nums
    
    
a = Solution()
print(a.rotate([1,2,3,4,5,6,7], 3)) # [5,6,7,1,2,3,4]
print(a.rotate([-1,-100,3,99], 2)) # [3,99,-1,-100]

File: 189_Rotate_Array.c
#include <stdio.h>
#include <stdlib.h>

void rotate(int* nums, int numsSize, int k) {
    int temp, start = 0, end = numsSize - 1;

    if (numsSize == 0 || k <= 0 ) {
        return;
    }
    k = k % numsSize; // Handle cases where k is greater than numsSize

    while (start < end) {
        temp = nums[start];
        nums[start] = nums[end];
        nums[end] = temp;
        start++;
        end--;
    }

    end = k-1, start = 0;
    while (start < end) {
        temp = nums[start];
        nums[start] = nums[end];
        nums[end] = temp;
        start++;
        end--;
    }   

    end = numsSize - 1;
    while (k < end) {
        temp = nums[k];
        nums[k] = nums[end];
        nums[end] = temp;
        k++;
        end--;
    }
}


int main() {
    int nums[] = {1, 2, 3, 4, 5, 6, 7};
    int k = 3;
    int numsSize = sizeof(nums) / sizeof(nums[0]);

    rotate(nums, numsSize, k);

    for (int i = 0; i < numsSize; i++) {
        printf("%d ", nums[i]);
    }
    printf("\n");

    return 0;
}